### PR TITLE
refactor: change format of generated typescript from resource store

### PR DESCRIPTION
**Pull request by jrowland-ccc**
Updates the typescript generation template to desired format.

Also refactors the methods associated with typescript template rendering: 

- Map types like `map[string]accesstypes.Resource` can be replaced with `map[accesstypes.Resource]struct{}` for reduced memory usage and simpler typing. 
- Removed conversions of `accesstypes.Resource` to `string` because the type is already an alias of `string`.  
- Removed `fmt.Sprintf("%s_%s", resource, tag)` occurrences to leverage `ResourceWithTag(tag)` method built into the `accesstypes.Resource` type. 
- Some maps were created with make or composite literals side-by-side, I tried to make it more uniform where I could. 
- I moved some duplicate loops into a nested function within `permissionResources` method, with the hope it improves readability, but it's not strictly necessary.

### CODE REVIEW

File: resourcestore/generate_typescript_test.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+			if diff := cmp.Diff(want, got); (diff != "") != tt.wantDiff {
+				t.Errorf("Store.GenerateTypeScript() output mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+func fakeStores(t *testing.T) fields {
+	t.Helper()
```
Comment (bswaney-ccc) [inline resourcestore/generate_typescript_test.go:74]: Nice attention to detail!

File: resourcestore/generate_typescript_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+			"addr": []accesstypes.Permission{"Create", "Delete"},
+		},
+		"Prototype3": {
+			"socket":  []accesstypes.Permission{},
+			"sockopt": []accesstypes.Permission{"Read", "List"},
+		},
+	}
+
+	return fields{map[accesstypes.PermissionScope]tagStore{"global": tStore}, map[accesstypes.PermissionScope]resourceStore{"global": rStore}}
+}
```
[the remark this answers is not in the document]
Comment (bswaney-ccc) [reply resourcestore/generate_typescript_test.go:99]: or even something like this
```go
func fakeStores(t *testing.T) fields {
	t.Helper()

	return fields{
		map[accesstypes.PermissionScope]tagStore{
			"global": {
				"Prototype1": {
					"id":       []accesstypes.Permission{"Create", "Delete"},
					"protocol": []accesstypes.Permission{"Create", "Delete"},
				},
				"Prototype2": {
					"id":   []accesstypes.Permission{"Create", "Delete"},
					"uuid": []accesstypes.Permission{"List", "Read", "Update", "Delete"},
					"addr": []accesstypes.Permission{"Create", "Delete"},
				},
				"Prototype3": {
					"socket":  []accesstypes.Permission{},
					"sockopt": []accesstypes.Permission{"Read", "List"},
				},
			},
		},
		map[accesstypes.PermissionScope]resourceStore{
			"global": {
				"Prototype1": {"Create", "Delete"},
				"Prototype2": {"Update", "List", "Read"},
				"Prototype3": {"Delete", "Read", "List"},
			},
		},
	}
}

```

File: resourcestore/generate_typescript_test.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	resourceStore map[accesstypes.PermissionScope]resourceStore
+}
+
+func TestStore_GenerateTypeScript(t *testing.T) {
+	type args struct {
+		dst string
+	}
+	tests := []struct {
+		name     string
+		fields   fields
```
Comment (bswaney-ccc) [inline resourcestore/generate_typescript_test.go:23]: We should add `t.Parallel()` to the top of this test.

File: resourcestore/generate_typescript.go [hunk 4]
```diff
@@ -52,17 +68,15 @@ func (s *Store) GenerateTypeScript(dst string) error {
...
-		Resources:   s.resources(),
-		Mappings:    s.permissionResources(),
+		Permissions:         s.permissions(),
+		Resources:           s.resources(),
+		ResourceTags:        s.tags(),
+		ResourcePermissions: s.resourcePermissions(),
 	}); err != nil {
-		panic(err)
+		return err
 	}
```
Comment (bswaney-ccc) [inline resourcestore/generate_typescript.go:80]: Our pattern is to wrap the errors that are coming out of external packages, with the function name. 

This is the typical pattern:
```go
	if err := tsFile.Execute(f, TSGenerator{
		Permissions:         s.permissions(),
		Resources:           s.resources(),
		ResourceTags:        s.tags(),
		ResourcePermissions: s.resourcePermissions(),
	}); err != nil {
		return errors.Wrap(err, "template.Template.Execute()")
	}
```

File: resourcestore/generate_typescript.go [hunk 5]
```diff
@@ -52,17 +68,15 @@ func (s *Store) GenerateTypeScript(dst string) error {
...
 
 	if err := f.Close(); err != nil {
```
Comment (bswaney-ccc) [inline resourcestore/generate_typescript.go:82]: Likewise these errors should be wrapped.

**Review by bswaney-ccc — CHANGES_REQUESTED** [review]
I am finding a lot of linter type errors in this PR, we should make sure that your linter is enable correctly. I requested a few of the changes, but the others will show up automatically, once we get it enabled.

File: resourcestore/generate_typescript.go [hunk 6]
```diff
@@ -52,21 +68,19 @@ func (s *Store) GenerateTypeScript(dst string) error {
...
-		panic(err)
+		return errors.Wrap(err, "template.Template.Execute()")
 	}
 
 	if err := f.Close(); err != nil {
-		return err
+		return errors.Wrap(err, "os.file.Close()")
 	}
 
 	return err
```
Comment (bswaney-ccc) [inline resourcestore/generate_typescript.go:86]: This err still needs wrapped

File: resourcestore/generate_typescript.go [hunk 7]
```diff
@@ -1,47 +1,63 @@
...
+	},
+		{{- end}}
+	{{- end}}
 };
-
-export function requiresPermission(resource: Resources, permission: Permissions): boolean {
-  return Mappings[resource][permission];
+export function requiresPermission(resource: AllResources, permission: Permissions): boolean {
+	return Mappings[resource][permission];
 }
```
Comment (bswaney-ccc) [inline resourcestore/generate_typescript.go:61]: I think we should add some whitespace between the major sections.

suggested:
```
{{- $permissions := .Permissions }}
{{- $resources := .Resources}}
{{- $resourcetags := .ResourceTags }}
{{- $resourcePerms := .ResourcePermissions}}
export enum Permissions {
{{- range $permissions}}
	{{.}} = '{{.}}',
{{- end}}
}

export enum Resources {
{{- range $resource := $resources}}
	{{$resource}} = '{{$resource}}',
{{- end}}
}

{{- range $resource, $tags := $resourcetags}}
export enum {{$resource}} {
	{{- range $_, $tag:= $tags}}
	{{$tag}} = '{{$resource.ResourceWithTag $tag}}',
	{{- end}}
}
{{- end}}

type AllResources = Resources {{- range $resource := .Resources}} | {{$resource}}{{- end}};
type PermissionResources = Record<Permissions, boolean>;
type PermissionMappings = Record<AllResources, PermissionResources>;

const Mappings: PermissionMappings = {
	{{- range $resource := $resources}}
	[Resources.{{$resource}}]: {
		{{- range $perm := $permissions}}
		[Permissions.{{$perm}}]: {{- index $resourcePerms $resource $perm}},
		{{- end}}
	},
		{{- range $tag := index $resourcetags $resource}}
	[{{$resource.ResourceWithTag $tag}}]: {
			{{- range $perm := $permissions}}
		[Permissions.{{$perm}}]: {{- index $resourcePerms ($resource.ResourceWithTag $tag) $perm}},
			{{- end}}
	},
		{{- end}}
	{{- end}}
};

export function requiresPermission(resource: AllResources, permission: Permissions): boolean {
	return Mappings[resource][permission];
}
```

**Review by jwatson-ccc — APPROVED** [review]
LGTM

**Review by bswaney-ccc — APPROVED** [review]
LGTM

**Pull request merged**